### PR TITLE
`Programming exercises`: Add possibility to import existing static code analysis configurations

### LINKED ISSUE #5169: Add coursewide default SCA configuration / allow to copy configurations
### Is your feature request related to a problem?

When activating Deductions for code stlye issues, the SCA grading needs to be configured. Usually, instructors want a consistent SCA configurations (and consistent deductions) among all exercises. This currenlty is only possible, by setting up the configuration for every exercise per hand. 

### Describe the solution you'd like

There are two different solutions I could think of:

1. You can specify a course-wide default grading. This gets copyd once a new programming exercise gets created. (We need a copy here to provide a common starting place while stil allowing instructors to do exercise-specific changes)

2. The grading-tab could feature a button "copy from exercise". This would also give the advantage to have mutliple standard configurations (e.g one for homeworks and one for tutor group exercises). 

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Strohgelaender**
# Note: This PR is based on #6026

#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
Closes #5169

Typically, Instructors want to use one standard SCA configuration for all exercises. This allows consistency in what type of feedback students get to see and how many points these issues deduct. Currently, setting up this configuration is cumbersome, since every category must be configured again and again for every exercise manually.

This PR introduces a new mechanism allowing editors to copy over an existing SCA configuration from another exercise. After defining this configuration once, you no longer need to define the same configuration again and again, but just import it into other exercises.

### Description

This PR introduces a new REST endpoint `/static-code-analysis-categories/import` which allows copying over the configuration from one exercise into another. The calling user must have at least editor rights for both exercises.

Additionally, this PR refactors the database calls when finding the available exercises on the different pages of the import dialog. Currently, for every possible filter combination and whether the current user is an admin or not, an own method in the different repositories exists (leading to 6 methods since the case of courseFilter == false and examFilter == false can be short-evaluated to just return an empty list). These methods were also duplicated in all 4 exercise repositories. With the addition of another `SCAFilter`. this approach was no longer maintainable. 

Instead, this PR introduces the [Criteria builder](https://docs.artemis.cit.tum.de/dev/guidelines/criteria-builder) which is already used for filtering posts. This allows us to dynamically create efficient queries reacting to the different filters. In order to reuse this implementation for all exercise types, a new `ExerciseSpecificationService` got added, which dynamically builds the specification based on the search string, the filters, and the settings (page number, page size, sorting).

### Steps for Testing
Prerequisites:
- 1 Instructor or Editor
- 2 Programming Exercises with SCA enabled
- 1 Programming Exercise with SCA deactivated

1. Setup one exercise to have a distinctive SCA configuration
2. Navigate to the grading page of the second exercise and press on the "Import Configuration" button
3. Check that only exercises with SCA active are available in the import dialog and that search / sort / and filters work.
4. Select one exercise and press import.
5. Verify that the other SCA categories got imported into the existing exercise.
6. Submit solutions as a Student and check that the configuration gets applied correctly.

#### Exam Mode Testing
Do the same steps as above. Verify that importing from and to exam exercises is possible.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test

### Test Coverage
100% coverage for changed / new lines.
| Class/File | Branch | Line |
|------------|-------:|-----:|
| StaticCodeAnalysisService.java | 97% | 91% |
| StaticCodeAnalysisRessource.java | 100% | 90% |
| ExerciseSpecificationService.java | 100% | 100% |
| ProgrammingExerciseResource.java | 91% | 83% |

### Screenshots
TODO

### CODE REVIEW

File: src/main/webapp/i18n/en/programmingExercise.json [hunk 1]
```diff
@@ -174,7 +174,11 @@
                     "updated": "The categories were updated successfully.",
                     "couldNotBeUpdated": "The following categories could not be updated: {{categories}}.",
                     "resetSuccessful": "The categories were reset successfully.",
-                    "resetFailed": "The categories could not be reset."
+                    "resetFailed": "The categories could not be reset.",
+                    "import": "Import Configuration",
+                    "importLabel": "Import SCA configuration from exercise",
+                    "importSuccessful": "Successfully imported the configuration from exercise {{exercise}}",
+                    "importFailed": "Failed to import categories from exercise {{exercise}}"
```
Comment (b-fein) [inline src/main/webapp/i18n/en/programmingExercise.json:181]:
```suggestion
                    "importSuccessful": "Successfully imported the configuration from exercise {{exercise}}.",
                    "importFailed": "Failed to import categories from exercise {{exercise}}."
```

File: src/test/java/de/tum/in/www1/artemis/util/ExerciseIntegrationTestUtils.java [hunk 2]
```diff
@@ -61,6 +63,49 @@ public void testCourseAndExamFilters(String apiPath, String searchTerm) throws E
...
+                var exerciseComparator = getExpetedComparator(sort);
+                if (order == SortingOrder.DESCENDING) {
+                    exerciseComparator = exerciseComparator.reversed();
+                }
+                assertThat(result.getResultsOnPage()).as("Sorting by " + sort + " " + order).isSortedAccordingTo(exerciseComparator);
+            }
+        }
+    }
+
+    private Comparator<LinkedHashMap<String, ? extends Comparable>> getExpetedComparator(Exercise.ExerciseSearchColumn sort) {
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/util/ExerciseIntegrationTestUtils.java:95]:
```suggestion
    private Comparator<LinkedHashMap<String, ? extends Comparable>> getExpectedComparator(Exercise.ExerciseSearchColumn sort) {
```

File: src/test/java/de/tum/in/www1/artemis/util/ExerciseIntegrationTestUtils.java [hunk 3]
```diff
@@ -61,6 +63,49 @@ public void testCourseAndExamFilters(String apiPath, String searchTerm) throws E
...
+                if (sort == Exercise.ExerciseSearchColumn.EXAM_TITLE) {
+                    params.add("isCourseFilter", "false");
+                }
+                else if (sort == Exercise.ExerciseSearchColumn.COURSE_TITLE) {
+                    params.add("isExamFilter", "false");
+                }
+
+                var result = request.get(apiPath, HttpStatus.OK, SearchResultPageDTO.class, params);
+
+                var exerciseComparator = getExpetedComparator(sort);
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/util/ExerciseIntegrationTestUtils.java:86]:
```suggestion
                var exerciseComparator = getExpectedComparator(sort);
```

**Review by tobias-lippert — APPROVED** [review]
Code LGTM. I like the refactoring :)

File: src/main/java/de/tum/in/www1/artemis/service/StaticCodeAnalysisService.java [hunk 4]
```diff
@@ -142,6 +142,32 @@ public Set<StaticCodeAnalysisCategory> resetCategories(ProgrammingExercise exerc
...
     }
 
+    /**
+     * This method allows users to reuse an already existing SCA configuration by copying it into another exercise.
+     * The previous configuration of the targeted exercise will get removed.
+     * @param targetExercise The exercise into which the configuration gets copied in
+     * @param sourceExercise The exercise to take the existing configuration from
+     * @return the new SCA configuration of the targetExercise
+     */
+    public Set<StaticCodeAnalysisCategory> importCategoriesFromExercise(ProgrammingExercise targetExercise, ProgrammingExercise sourceExercise) {
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/StaticCodeAnalysisService.java:152]:
```suggestion
    public Set<StaticCodeAnalysisCategory> importCategoriesFromExercise(ProgrammingExercise sourceExercise, ProgrammingExercise targetExercise) {
```
Personal preference: Copying from source to target and therefore specifying the parameters in that order would be more intuitive to me. Can of course be left as-is, though.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 5]
```diff
@@ -104,6 +103,8 @@ public class ProgrammingExerciseService {
 
     private final ProgrammingExerciseGitDiffReportRepository programmingExerciseGitDiffReportRepository;
 
+    private final ExerciseSpecificationService exerciseSpecificationService;
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:106]: `authCheckService` attribute above is now unused.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 6]
```diff
@@ -951,31 +954,40 @@ public ProgrammingExercise save(ProgrammingExercise programmingExercise) {
      */
     public SearchResultPageDTO<ProgrammingExercise> getAllOnPageWithSize(final PageableSearchDTO<String> search, final Boolean isCourseFilter, final Boolean isExamFilter,
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:955]:
```suggestion
    public SearchResultPageDTO<ProgrammingExercise> getAllOnPageWithSize(final PageableSearchDTO<String> search, final boolean isCourseFilter, final boolean isExamFilter,
```
Can you change the parameter types here, or would that require some refactoring in other places?
Applies to equivalent method on services for other exercise types as well.

**Review by b-fein — APPROVED** [review]
Code changes look good.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java (hunk 6 above)
Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:955]: They are annotated with `@RequestParam(defaultValue = "true")` so null should not be possible.

**Review by b-fein — APPROVED** [review]
Re-review: Code changes look good. :+1:

**Review by tobias-lippert — APPROVED** [review]
Reapprove code

**Review by kekskatze — APPROVED** [review]
Tested on ts1. Everything worked as expected

**Review by luisa1999 — APPROVED** [review]
Used ts2 for testing and everything worked.

**Review by frederik-kt — APPROVED** [review]
Tested on TS1. Importing the SCA configuration worked as expected.

**Review by danielbode — APPROVED** [review]
Tested on ts1. Works as expected.

File: src/main/webapp/app/exercises/shared/import/exercise-import.component.ts [hunk 7]
```diff
@@ -71,6 +83,9 @@ export class ExerciseImportComponent implements OnInit {
             case ExerciseType.MODELING:
                 return this.injector.get(ModelingExercisePagingService);
             case ExerciseType.PROGRAMMING:
+                if (this.programmingLanguage) {
```
[the remark this answers is not in the document]
Comment (Strohgelaender) [reply src/main/webapp/app/exercises/shared/import/exercise-import.component.ts:86]: You mean that `programmingLanguage` is undefined? Yes, this is expected to happen. It's only set if we want to filter for one language.

**Review by Vinicius-Agreste — APPROVED** [review]
Tested on ts1. Worked as expected.

**Review by simon-hng — APPROVED** [review]
Code

**Pull request merged**